File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/application/core/usecase/ListValorantAccountsUseCase.java
package fr.huiitre.tools.modules.riot.valorant.application.core.usecase;

import fr.huiitre.tools.modules.core.module.domain.ModuleCode;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.ports.AuthenticatedUserProvider;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.core.ports.ValorantAuthRepository;
import fr.huiitre.tools.modules.riot.valorant.application.core.view.ValorantAccountView;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class ListValorantAccountsUseCase implements SecuredUseCase {

    private final ValorantAuthRepository valorantAuthRepository;
    private final AuthenticatedUserProvider authenticatedUserProvider;

    public ListValorantAccountsUseCase(ValorantAuthRepository valorantAuthRepository,
                                        AuthenticatedUserProvider authenticatedUserProvider) {
        this.valorantAuthRepository = valorantAuthRepository;
        this.authenticatedUserProvider = authenticatedUserProvider;
    }

    @Override
    public Optional<ModuleCode> requiredModule() {
        return Optional.of(ModuleCode.RIOT);
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.READ_ONLY;
    }

    public List<ValorantAccountView> execute() {
        Long userId = authenticatedUserProvider.getUserId();
        return valorantAuthRepository.findAllByUserId(userId).stream()
                .map(a -> new ValorantAccountView(a.id(), a.puuid(), a.region(), a.gameName(), a.tagLine(), a.label()))
                .toList();
    }
}

File: api/Modules/Riot/Valorant/Application/Core/Usecases/ListValorantAccountsUseCase.cs
using Tools.Api.Modules.Core.Security.Application.Services;
using Tools.Api.Modules.Core.Security.Application.Usecases;
using Tools.Api.Modules.Core.Security.Domain;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Views;

namespace Tools.Api.Modules.Riot.Valorant.Application.Core.Usecases;

public sealed class ListValorantAccountsUseCase(
    UseCaseAuthorizer authorizer,
    IValorantAuthRepository valorantAuthRepository
) : SecuredUseCase(authorizer)
{
    protected override RoleCode RequiredRole => RoleCode.ReadOnly;
    protected override ModuleCode? RequiredModule => ModuleCode.Riot;

    // Le propriétaire n'est pas un argument : c'est l'appelant validé.
    public async Task<List<ValorantAccountView>> Execute()
    {
        var accounts = await valorantAuthRepository.FindAllByUserId(CurrentUser.UserId);

        return accounts
            .Select(account => new ValorantAccountView(
                account.Id,
                account.Puuid,
                account.Region,
                account.GameName,
                account.TagLine,
                account.Label))
            .ToList();
    }
}
